### PR TITLE
Add better guard against inserting item from other list into linked list

### PR DESCRIPTION
**Pull request by bantic**
Adds `__parent` property to items to identify their inclusion in a list.
This allows us to always know when we are inserting an item from another
list. The previous heuristic (checking for presence of `next` and `prev`) is
not correct in all scenarios, leading to unusual bugs.

This refactors `insertBefore` method slightly for clarity, and
simplifies `remove` because we now know for sure when we are removing
an item in this list.

Adds some additional test coverage.

### CODE REVIEW

File: src/js/utils/linked-list.js [hunk 1]
```diff
@@ -11,9 +15,15 @@ export default class LinkedList {
...
   }
   adoptItem(item) {
+    item[PARENT_PROP]= this;
+    this.length++;
     if (this._adoptItem) { this._adoptItem(item); }
   }
   freeItem(item) {
+    item[PARENT_PROP] = null;
+    item.prev   = null;
+    item.next   = null;
```
Comment (mixonic) [inline src/js/utils/linked-list.js]: Shouldn't this already be done by `remove`?

Comment (mixonic) [inline src/js/utils/linked-list.js]: Ah you've moved it. As `prev` and `next` are related to position and not ownerships (`adopt`/`free`) IMO the changes should be made in remove.

**Comment by mixonic** [comment]
:+1:

File: src/js/utils/linked-list.js (hunk 1 above)
Comment (bantic) [inline src/js/utils/linked-list.js]: Makes sense. I've moved it. Thanks for looking it over.

**Pull request merged**